### PR TITLE
refactor: roll RegionedWritePass into WritePass

### PR DESCRIPTION
**Pull request by rphmeier**

### CODE REVIEW

**Comment by rphmeier** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/171?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#171** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/171?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#166** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/166?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#170](https://github.com/thrumdev/nomt/pull/170) <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/170?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#165** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/165?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#164** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/164?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#161** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/161?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rphmeier and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: nomt/src/rw_pass_cell.rs [hunk 1]
```diff
@@ -90,61 +90,213 @@ impl RwPassDomain {
...
+        } else {
+            None
+        }
+    }
 }
 
+impl WritePass<UniversalRegion> {
+    /// Supply a region type.
+    pub fn with_region<R>(self, region: R) -> WritePass<R> {
+        // sanity: UniversalRegion can't be split, so this should never be Some.
```
Comment (gabriele-0201) [inline nomt/src/rw_pass_cell.rs:249]: The implementation of `RegionContains` returns true for both `contains` and `contains_exclusive`. Why `split` should fail if called on `WritePass<UniversalRegion>` ? The two asserts seems to be satisfied at the beginning of `split`

File: nomt/src/rw_pass_cell.rs [hunk 2]
```diff
@@ -173,8 +325,14 @@ impl<T, Id> RwPassCell<T, Id> {
     /// Returns a handle to read the value. Requires showing a read pass.
     ///
     /// Panics if the provided pass belongs to a different [`RwPassDomain`] than the cell.
-    pub fn read<'a, 'pass>(&'a self, read_pass: &'pass ReadPass) -> ReadGuard<'a, 'pass, T> {
+    /// If this pass is constrained to a specific region, this panics if the item's ID is outside
+    /// of the region's exclusive access range.
```
Comment (gabriele-0201) [inline nomt/src/rw_pass_cell.rs:329]: "exclusive *or not* access range"

Should the read pass be able to access exclusive and non-exclusive ids?

File: nomt/src/rw_pass_cell.rs [hunk 3]
```diff
@@ -322,41 +447,35 @@ impl<'a, 'pass, T> DerefMut for WriteGuard<'a, 'pass, T> {
...
-pub enum RegionContains {
-    /// Read-only (shared) access.
-    Read,
-    /// Read and Write (exclusive) access.
-    ReadWrite,
-}
-
-/// `Region`s indicate a range of data identifiers. This trait is unsafe because the memory safety
-/// of [`RegionedWritePass`] depends on the results of its functions.
+/// Whether a region contains values of not.
```
Comment (gabriele-0201) [inline nomt/src/rw_pass_cell.rs:450]: Whether a region contains values **or** not ?

**Review by gabriele-0201 — APPROVED** [review]

File: nomt/src/rw_pass_cell.rs [hunk 4]
```diff
@@ -322,41 +447,35 @@ impl<'a, 'pass, T> DerefMut for WriteGuard<'a, 'pass, T> {
...
     ///
-    /// This region must encompass all other regions, exclude all other regions, and contain
-    /// all values for write access.
-    fn universe() -> Self;
+    /// If this function is ever invoked with an ID, it must return the same result whenever it is
+    /// called in the future.
+    fn contains(&self, id: &Id) -> bool;
 
-    /// Whether the region contains a value.
+    /// Whether the region contains a ID exclusively..
```
Comment (pepyakin) [inline nomt/src/rw_pass_cell.rs:468]:
```suggestion
    /// Whether the region contains a ID exclusively.
```

File: nomt/src/rw_pass_cell.rs [hunk 5]
```diff
@@ -322,41 +447,35 @@ impl<'a, 'pass, T> DerefMut for WriteGuard<'a, 'pass, T> {
...
     ///
     /// # Safety
     ///
     /// If this function is ever invoked with an ID, it must return the same result whenever it is
     /// called in the future.
-    fn contains(&self, value: &Self::Id) -> Option<RegionContains>;
+    fn contains_exclusive(&self, id: &Id) -> bool;
+}
 
+/// `Region`s indicate a range of data identifiers.
```
Comment (pepyakin) [inline nomt/src/rw_pass_cell.rs:477]: Now that `Region` doesn't have a notion of an `Id` this kinda sounds confusing.

**Review by pepyakin — APPROVED** [review]

**Pull request merged**